### PR TITLE
symmetry bug fix

### PR DESCRIPTION
**Pull request by sunghjung3**
Fixed bug related to the symmetry of 3-body interactions.

The determination of the `self.symmetry` variable was moved out of `update_knots()` and as its own method called `find_symmetry_3B()`. It now looks at the identities of the 2 neighbor atoms to determine the symmetry, in addition to the knot settings.

The 3-body coefficient tensor compression and decompression algorithms were also slightly modified. Previously, all compression and decompression of 3-body terms were done assuming that it had 2-way symmetry. However, now it can handle 1-way and 3-way symmetries too.

### CODE REVIEW

**Comment by monk-04** [comment]
Thank you for finding this bug.

I don't think the symmetry finding code should be outside of the `update_knots()` function.

In the current form `find_symmetry_3B()` accepts a three body interaction tuple (`trio`) as an input argument and sets the value for `self.symmetry[trio]` variable. Because the way it is written, an user can accidentally add unnecessary 3-body interaction to the `self.symmetry`. 

For example,  for a binary system 'Pt-C', I can add arbitrary 3-body interactions like shown below-
image


I propose the following changes instead-

```
            min_set = len(set(self.r_min_map[trio]))                            
            max_set = len(set(self.r_max_map[trio]))                            
            res_set = len(set(self.resolution_map[trio]))                       
            if trio[1]!=trio[2]:                                                
                self.symmetry[trio] = 1                                         
            else:                                                               
                if min_set == 1 and max_set == 1 and res_set == 1:              
                    self.symmetry[trio] = 3                                     
                elif min_set <= 2 and max_set <= 2 and res_set <= 2:            
                    self.symmetry[trio] = 2                                     
                else:                                                           
                    self.symmetry[trio] = 1 


```

The proposed changes only adds an if-else statement leaving the original code in-place.

**Comment by sunghjung3** [comment]
I guess I was unconsciously fixing yet another bug here. Those if-else statements with `min_set`, `max_set`, and `res_set` are not correct. For example, if a 3-body interaction `trio` has `self.r_min_map[trio] = [1, 1, 2]`, `self.r_max_map[trio] = [5, 6, 5]`, and `self.resolution_map[trio] = [10, 20, 20]`, we would get `min_set == max_set == res_set == 2`, although there is clearly no symmetry here because the knot intervals (assuming uniform spacing) would be `[0.4, 0.25, 0.15]`. My proposed edit in the `find_symmetry_3B()` method fixes this issue.

Since you want the fix to not have a new method in the class, I suggest that we copy the contents of the `find_symmetry_3B()` into the `update_knots()` method:

        # 2 neighboring elements are different
        if trio[1] != trio[2]:
            self.symmetry[trio] = 1
        else:  # 2 neighboring elements are identical
            configs = set(zip(
                self.r_min_map[trio],
                self.r_max_map[trio],
                self.resolution_map[trio]
            ))

            # 1->3, 2->2, 3->1
            self.symmetry[trio] = -1 * (len(configs) - 2) + 2

**Comment by sunghjung3** [comment]
@JasonGibsonUfl Thanks for the suggestion. I believe that the proposal is incorrect in 2 ways.

First, with `all(descriptor_sym_set)`, I think you meant to see if all the entries in the list `descriptor_sym_set` had the same values. However, I believe that it just sees that all entries are non-zero. These both return true:

    all([1, 1, 1]) ==> True
    all([1, 2, 3]) ==> True

Perhaps you meant to do this instead (although it does become a bit wordy):

    all(n == descriptor_sym_set[0] for n in descriptor_sym_set)

Even if this replacement is made, it becomes identical to @monk-04 's proposal [here](https://github.com/uf3/uf3/pull/60#issuecomment-1705688117), which does not fix the problem I outlined in my reply [here](https://github.com/uf3/uf3/pull/60#issuecomment-1705698247).

To clarify further about my proposed fix in the linked reply, the only line that could be confusing is this line:

            configs = set(zip(
            self.r_min_map[trio],
            self.r_max_map[trio],
            self.resolution_map[trio]
        ))

This just creates a set of `(r_min, r_max, resolution)` tuples for each dimension in the 3-body interaction. For instance:

    interaction: ('N', 'N', 'N')
    r_min_map: [a, b, c]
    r_max_map: [d, e, f]
    resolution_map: [g, h, I]
    ---
    configs == {(a, d, g), (b, e, h), (c, f, l)}

And the following line (copied below) sees how many unique elements are in this `configs` set and maps 1 to 3, 2 to 2, and 3 to 1:

        # 1->3, 2->2, 3->1
        self.symmetry[trio] = -1 * (len(configs) - 2) + 2

**Comment by sunghjung3** [comment]
Instead of writing `self.symmetry[trio] = -1 * (len(configs) - 2) + 2` in the last line, here is the expanded version for clarity:

    # 2 neighboring elements are different
    if trio[1] != trio[2]:
        self.symmetry[trio] = 1
    else:  # 2 neighboring elements are identical

        configs = list(zip(
            self.r_min_map[trio],
            self.r_max_map[trio],
            self.resolution_map[trio]
        ))
        central_legs_same = (configs[0] == configs[1])
        all_legs_same = (central_legs_same and configs[0] == configs[2])

        if all_legs_same: 
            self.symmetry[trio] = 3
        elif central_legs_same:
            self.symmetry[trio] = 2
        else:
            # 3 legs all have different configurations
            # OR the 2 central legs have different configurations
            self.symmetry[trio] = 1

I have also considered the case where the 1st (center to 1st neighbor) and 3rd (neighbor to neighbor) legs have the same knot configurations while the 2nd (center to 2nd neighbor) leg has a different configuration. In this case, I believe that the symmetry should be 2 since the [`compress_3B()`](https://github.com/uf3/uf3/blob/master/uf3/representation/bspline.py#L552) and the [`decompress_3B()`](https://github.com/uf3/uf3/blob/master/uf3/representation/bspline.py#L579C16-L579C16) functions also interpret `self.symmetry[interaction] == 2` to be the symmetry between the 2 central legs.

Let me know any thoughts about this very last point.

**Comment by monk-04** [comment]
I like the changes that @JasonGibsonUfl  made. Just waiting for him to push his unit tests.

**Comment by sunghjung3** [comment]
@JasonGibsonUfl Done.

**Comment by monk-04** [comment]
There's no documentation for bspline_config.compress_3B and bspline_config.decompress_3B. I will open another pull request to add it.
